tools-local/tasks/mobile.tasks/WasmAppBuilder/IcallTableGenerator.cs

```diff
@@ -0,0 +1,301 @@
+        foreach (var icall in _icalls)
+            assemblyMap [icall.Assembly!] = icall.Assembly!;
+
+        foreach (var assembly in assemblyMap.Keys)
+        {
+            var sorted = _icalls.Where (i => i.Assembly == assembly).ToArray ();
+            Array.Sort (sorted);
+
+            string aname;
+            if (assembly == "mscorlib" || assembly == "System.Private.CoreLib")
+                aname = "corlib";
+            else
+                aname = assembly.Replace (".", "_");
+            w.WriteLine ($"#define ICALL_TABLE_{aname} 1\n");
+
+            w.WriteLine ($"static int {aname}_icall_indexes [] = {{");
+            foreach (var icall in sorted)
+                w.WriteLine (string.Format ("{0},", icall.TokenIndex));
+            w.WriteLine ("};");
+            foreach (var icall in sorted)
+                w.WriteLine (GenIcallDecl (icall));
+            w.WriteLine ($"static void *{aname}_icall_funcs [] = {{");
+            foreach (var icall in sorted)
+            {
+                w.WriteLine (string.Format ("// token {0},", icall.TokenIndex));
+                w.WriteLine (string.Format ("{0},", icall.Func));
+            }
+            w.WriteLine ("};");
+            w.WriteLine ($"static uint8_t {aname}_icall_handles [] = {{");
+            foreach (var icall in sorted)
+                w.WriteLine (string.Format ("{0},", icall.Handles ? "1" : "0"));
+            w.WriteLine ("};");
+        }
+    }
+
+    // Read the icall table generated by mono --print-icall-table
+    private void ReadTable (string filename)
+    {
+        JsonDocument json;
+        using (var stream = File.Open (filename, FileMode.Open))
+        {
+            json = JsonDocument.Parse (stream);
+        }
+
+        var arr = json.RootElement;
+        foreach (var v in arr.EnumerateArray ())
+        {
+            var className = v.GetProperty ("klass").GetString ()!;
+            if (className == "")
+                // Dummy value
+                continue;
+            var icallClass = new IcallClass (className);
+            _runtimeIcalls [icallClass.Name] = icallClass;
+            foreach (var icall_j in v.GetProperty ("icalls").EnumerateArray ())
+            {
+                if (!icall_j.TryGetProperty ("name", out var nameElem))
+                    continue;
+                string name = nameElem.GetString ()!;
+                string func = icall_j.GetProperty ("func").GetString ()!;
+                bool handles = icall_j.GetProperty ("handles").GetBoolean ();
+
+                icallClass.Icalls [name] = new Icall (name, func, handles);
+            }
+        }
+    }
+
+    private void ProcessType (Type type)
+    {
+        foreach (var method in type.GetMethods(BindingFlags.DeclaredOnly|BindingFlags.Public|BindingFlags.NonPublic|BindingFlags.Static|BindingFlags.Instance))
+        {
+            if ((method.GetMethodImplementationFlags() & MethodImplAttributes.InternalCall) == 0)
```

Comment:
Should we move this to linker in the same way as we did for pinvoke export?

Reply:
We could do this, but the emitted information needs to include the signature because we have to map it to a C signature.

Reply:
Also, the current pinvoke table export in the linker doesn't include enough information for wasm either.

Reply:
What is missing?

Reply:
The signature, i.e. parameter/return types.

Reply:
I agree we should move this to the linker ideally.